common/functions/src/scalars/hashes/city64_with_seed.rs

```diff
@@ -0,0 +1,170 @@
+// Keep the struct within this file and don't expose outside.
+struct CityHasher64 {
+    seed: u64,
+    value: u64,
+}
+
+impl CityHasher64 {
+    fn with_seed(s: u64) -> Self {
+        Self { seed: s, value: 0 }
+    }
+}
+
+impl Hasher for CityHasher64 {
+    fn finish(&self) -> u64 {
+        self.value
+    }
+
+    fn write(&mut self, bytes: &[u8]) {
+        self.value = cityhash64_with_seed(bytes, self.seed);
+    }
+}
+
+#[derive(Clone)]
+pub struct City64WithSeedFunction {
+    display_name: String,
+}
+
+// CityHash64WithSeed(value, seed)
+impl City64WithSeedFunction {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function2>> {
+        Ok(Box::new(City64WithSeedFunction {
+            display_name: display_name.to_string(),
+        }))
+    }
+
+    pub fn desc() -> Function2Description {
+        Function2Description::creator(Box::new(Self::try_create))
+            .features(FunctionFeatures::default().deterministic().num_arguments(2))
+    }
+}
+
+impl Function2 for City64WithSeedFunction {
+    fn name(&self) -> &str {
+        &*self.display_name
+    }
+
+    fn get_monotonicity(
+        &self,
+        _args: &[crate::scalars::Monotonicity2],
+    ) -> Result<crate::scalars::Monotonicity2> {
+        Ok(crate::scalars::Monotonicity2::default())
+    }
+
+    fn return_type(
+        &self,
+        args: &[&common_datavalues2::DataTypePtr],
+    ) -> Result<common_datavalues2::DataTypePtr> {
+        if !matches!(
+            args[0].data_type_id(),
+            TypeID::UInt8
+                | TypeID::UInt16
+                | TypeID::UInt32
+                | TypeID::UInt64
+                | TypeID::Int8
+                | TypeID::Int16
+                | TypeID::Int32
+                | TypeID::Int64
+                | TypeID::Float32
+                | TypeID::Float64
+                | TypeID::Date16
+                | TypeID::Date32
+                | TypeID::DateTime32
+                | TypeID::DateTime64
+                | TypeID::Interval
+                | TypeID::String
+        ) {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Unsupported data type: {:?}",
+                args[0]
+            )));
+        }
+
+        if !args[1].data_type_id().is_numeric() {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Unsupported data type of hash seed: {:?}",
+                args[1]
+            )));
+        }
+        Ok(UInt64Type::arc())
+    }
+
+    fn eval(
+        &self,
+        columns: &common_datavalues2::ColumnsWithField,
+        _input_rows: usize,
+    ) -> Result<common_datavalues2::ColumnRef> {
+        let column = columns[0].column();
+        let physical_data_type = columns[0].data_type().data_type_id().to_physical_type();
+
+        let const_col: Result<&ConstColumn> = Series::check_get(columns[1].column());
+
+        if let Ok(col) = const_col {
```

Comment:
I think we can use `ScalarBinaryExpression` which covers the constant cases. And we do not need `cast_column_field` if we add bound trait: `AsPrimitive<u64>`

Reply:
For the constant cast, we can keep re-using the same hasher object because they have the same seed; non-constant case will need to create hasher per row.
ScalarBinaryExpression is good for this case, but I didn't figure out a way to avoid creating duplicate hasher.